File: leetcode/004_median_of_two_sorted_arrays.go
package main

import "fmt"

func findKth(a, b []int, k int) float64 {
	if len(a) == 0 {
		return float64(b[k-1])
	}
	if len(b) == 0 {
		return float64(a[k-1])
	}
	if k == 1 {
		if a[0] < b[0] {
			return float64(a[0])
		}
		return float64(b[0])
	}
	al := k / 2
	bl := k / 2
	if len(a) < al {
		al = len(a)
	}
	am := a[al-1]
	if len(b) < bl {
		bl = len(b)
	}
	bm := b[bl-1]
	if am > bm {
		return findKth(a, b[bl:], k-bl)
	}
	if am < bm {
		return findKth(a[al:], b, k-al)
	}
	if k-al-bl > 0 {
		return findKth(a[al:], b[bl:], k-al-bl)
	}
	return float64(am)
}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	l := len(nums1) + len(nums2)
	if l%2 == 0 {
		m1 := findKth(nums1, nums2, l/2)
		m2 := findKth(nums1, nums2, l/2+1)
		return (m1 + m2) / 2
	}
	return findKth(nums1, nums2, l/2+1)
}

func main() {
	n1 := []int{1, 2}
	n2 := []int{1, 4}
	fmt.Println(findMedianSortedArrays(n1, n2))
}

File: leetcode/004_median_of_two_sorted_arrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        l = len(nums1) + len(nums2)
        if l % 2 == 0:
            m1 = self.findKth(nums1, nums2, l/2)
            m2 = self.findKth(nums1, nums2, l/2+1)
            return (m1+m2)/2
        return self.findKth(nums1, nums2, l/2+1)

    def findKth(self, nums1, nums2, k):
        if len(nums1) == 0:
            return float(nums2[k-1])
        if len(nums2) == 0:
            return float(nums1[k-1])
        if k == 1:
            if nums1[0] < nums2[0]:
                return float(nums1[0])
            return float(nums2[0])
        al = k/2
        bl = al
        if len(nums1) < al:
            al = len(nums1)
        am = nums1[al-1]
        if len(nums2) < bl:
            bl = len(nums2)
        bm = nums2[bl-1]
        if am > bm:
            return self.findKth(nums1, nums2[bl:], k-bl)
        if am < bm:
            return self.findKth(nums1[al:], nums2, k-al)
        if k-al-bl>0:
            return self.findKth(nums1[al:], nums2[bl:], k-al-bl)
        return float(am)

s = Solution()
n1 = []
n2 = [2,3]
print s.findMedianSortedArrays(n1, n2)
